Fix thread race conditions when automatically registering components with Castle Windsor

MattFenelon opened the pull request:
Hi Seb,

I'm hoping this is a fix for the race conditions mentioned in [this thread](http://groups.google.com/group/openrasta/browse_thread/thread/6ac8f36beab63ab6?pli=1) on the OR mailing list. We've also seen the issue crop up when registering codecs.

I'd be interested to hear your thoughts on whether there would be any problems registering this way?

MattFenelon committed:
For some componennts OR will register the components with the IoC container just-in-time. This can cause race conditions if many requests are attempting to register the component at the same time. Added locking when adding dependencies to Windsor's container to fix this.

serialseb commented:
This would assume that we are happy to have the IDependencyResolver be designed as thread-safe, which means changing all the other container implementations. If it's a castle-specific problem where the methods on IWindsorContainer is not thread safe, then that patch can be fine. Can someone check with them?

The issue that's higlighted in that thread cannot happen for codecs, those are registered upon startup, once only and the configuration itself is thread safe. If you're seeing codec-related issues, it's not the same problem.

MattFenelon commented:
It does look like we are suffering the same problem, with our codecs, as spoken about in the original thread. In the stack trace below, the exception originates in `ResponseEntityWriterContributor.WriteResponse(ICommunicationContext context)`, when it is calling `DependencyManager.GetService`. The only usage I can find of GetService, inside that method, is when the `ResponseCodec.CodecType` is resolved. The resolution of the codec fails and falls through to the auto registration process.

```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: capacity
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument)
   at System.Collections.Generic.Dictionary`2..ctor(Int32 capacity, IEqualityComparer`1 comparer)
   at System.Collections.Generic.Dictionary`2..ctor(IDictionary`2 dictionary, IEqualityComparer`1 comparer)
   at System.Collections.Generic.Dictionary`2..ctor(IDictionary`2 dictionary)
   at Castle.MicroKernel.Handlers.AbstractHandler.DependencySatisfied(Boolean& stateChanged)
   at Castle.MicroKernel.DefaultKernel.DoActualRaisingOfHandlersChanged()
   at Castle.MicroKernel.DefaultKernel.RaiseHandlersChanged()
   at Castle.MicroKernel.DefaultKernel.RegisterHandler(String key, IHandler handler, Boolean skipRegistration)
   at Castle.MicroKernel.DefaultKernel.RegisterHandler(String key, IHandler handler)
   at Castle.MicroKernel.DefaultKernel.AddComponent(String key, Type serviceType, Type classType, LifestyleType lifestyle, Boolean overwriteLifestyle)
   at Castle.Windsor.WindsorContainer.AddComponentLifeStyle(String key, Type serviceType, Type classType, LifestyleType lifestyle)
   at OpenRasta.DI.Windsor.WindsorDependencyResolver.AddDependencyCore(Type dependent, Type concrete, DependencyLifetime lifetime)
   at OpenRasta.DI.Windsor.WindsorDependencyResolver.AddDependencyCore(Type handlerType, DependencyLifetime lifetime)
   at OpenRasta.DI.DependencyResolverCore.AddDependency(Type concreteType, DependencyLifetime lifetime)
   at OpenRasta.DI.DependencyManager.GetService(Type dependencyType)
   at OpenRasta.Pipeline.Contributors.ResponseEntityWriterContributor.WriteResponse(ICommunicationContext context)
   at OpenRasta.Pipeline.PipelineRunner.ExecuteContributor(ICommunicationContext context, ContributorCall call)
```

I think the reason we, specifically, are seeing this problem, is due to our use of custom codecs, which are registered using the fluent interface. Searching through the OR code I found that the default codecs are registered, with the IoC container, at application start up. I've been unable to find any code that registers codecs configured using the FI, with the IoC container, other than the just-in-time statements, as used in `ResponseEntityWriterContributor` above. Does this seem right, or have I missed something?

serialseb commented:
Ah. If that's the case that's a bug :) tbh auto registration ought to be removed and enforce reg on startup for both handlers and codecs, I suggest we apply that fix as a more long term solution, what do you think?

On 25 May 2011, at 15:29, "MattFenelon" reply@reply.github.com wrote:

> It does look like we are suffering the same problem, with our codecs, as spoken about in the original thread. In the stack trace below, the exception originates in `ResponseEntityWriterContributor.WriteResponse(ICommunicationContext context)`, when it is calling `DependencyManager.GetService`. The only usage I can find of GetService, inside that method, is when the `ResponseCodec.CodecType` is resolved. The resolution of the codec fails and falls through to the auto registration process.
> 
> ```
> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
> Parameter name: capacity
>   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument)
>   at System.Collections.Generic.Dictionary`2..ctor(Int32 capacity, IEqualityComparer`1 comparer)
>   at System.Collections.Generic.Dictionary`2..ctor(IDictionary`2 dictionary, IEqualityComparer`1 comparer)
>   at System.Collections.Generic.Dictionary`2..ctor(IDictionary`2 dictionary)
>   at Castle.MicroKernel.Handlers.AbstractHandler.DependencySatisfied(Boolean& stateChanged)
>   at Castle.MicroKernel.DefaultKernel.DoActualRaisingOfHandlersChanged()
>   at Castle.MicroKernel.DefaultKernel.RaiseHandlersChanged()
>   at Castle.MicroKernel.DefaultKernel.RegisterHandler(String key, IHandler handler, Boolean skipRegistration)
>   at Castle.MicroKernel.DefaultKernel.RegisterHandler(String key, IHandler handler)
>   at Castle.MicroKernel.DefaultKernel.AddComponent(String key, Type serviceType, Type classType, LifestyleType lifestyle, Boolean overwriteLifestyle)
>   at Castle.Windsor.WindsorContainer.AddComponentLifeStyle(String key, Type serviceType, Type classType, LifestyleType lifestyle)
>   at OpenRasta.DI.Windsor.WindsorDependencyResolver.AddDependencyCore(Type dependent, Type concrete, DependencyLifetime lifetime)
>   at OpenRasta.DI.Windsor.WindsorDependencyResolver.AddDependencyCore(Type handlerType, DependencyLifetime lifetime)
>   at OpenRasta.DI.DependencyResolverCore.AddDependency(Type concreteType, DependencyLifetime lifetime)
>   at OpenRasta.DI.DependencyManager.GetService(Type dependencyType)
>   at OpenRasta.Pipeline.Contributors.ResponseEntityWriterContributor.WriteResponse(ICommunicationContext context)
>   at OpenRasta.Pipeline.PipelineRunner.ExecuteContributor(ICommunicationContext context, ContributorCall call)
> ```
> 
> I think the reason we, specifically, are seeing this problem, is due to our use of custom codecs, which are registered using the fluent interface. Searching through the OR code I found that the default codecs are registered, with the IoC container, at application start up. I've been unable to find any code that registers codecs configured using the FI, with the IoC container, other than the just-in-time statements, as used in `ResponseEntityWriterContributor` above. Does this seem right, or have I missed something?
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/openrasta/openrasta-stable/pull/16#issuecomment-1235603

MattFenelon commented:
I second that. When the auto-registration problem bit us again, this time with the codecs (handlers got us first), I wanted to put something more permanent in place than registering these components, manually, at app startup. Hence the locking. Removing auto-reg completely would put my mind at rest :)

It does sound like a major change though? It'd probably be worth while to put this small patch in place, in the meantime, to stop others suffering from the same issue.

MattFenelon commented:
Seb, in case you're interested I've started a thread about this problem on the Castle mailing list:
http://groups.google.com/group/castle-project-users/browse_thread/thread/6f6edd7718c411c

MattFenelon committed:
_windsorContainer.AddComponentLifeStyle() is not thread safe, but Register() is.
Removed changes for 1.0 as I have no way of testing them.
Adding locking to AddDependencyInstanceCore as it can cause the container to get into an invalid state, if many threads are registering at the same time.
The lock object is now static to ensure it will apply across all instances of the WindsorDependencyResolver class.

MattFenelon commented:
I've updated the pull request because even after applying the original patch we continued to see similar issues with registration.

While investigating I found that the AddComponentLifeStyle methods on WindsorContainer are not thread-safe, nor are the methods on IKernel for registration. I've swaped those methods for their thread safe equivalents, Register(Component.For...) and removed the lock I put in place. I've also added more locking around the other routes into component registration.

MattFenelon committed:
Turns out the Register methods aren't thread-safe either.
